pkg/upgrade/v1alpha1/engine.go

```diff
@@ -0,0 +1,147 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"github.com/pkg/errors"
+
+	upgrade "github.com/openebs/maya/pkg/apis/openebs.io/upgrade/v1alpha1"
+	apis "github.com/openebs/maya/pkg/apis/openebs.io/v1alpha1"
+	stringer "github.com/openebs/maya/pkg/apis/stringer/v1alpha1"
+	cast "github.com/openebs/maya/pkg/castemplate/v1alpha1"
+)
+
+const (
+	// upgradeItemProperty Property used to store task configuration
+	// to get these value use {{ .UpgradeItem.<key> }} in runtask
+	upgradeItemProperty string = "UpgradeItem"
+	// configProperty Property used to store default configuration
+	// present in castemplate
+	// to get these value use {{ .Config.<key>.value }} in runtask
+	configProperty string = "Config"
+	// runtimeProperty Property used to store runtime configuration
+	// to get these value use {{ .Runtime.<key>.value }} in runtask
+	runtimeProperty string = "Runtime"
+	// nameKey is a property of TaskConfig this is name of the resource.
+	nameKey string = "name"
+	// namespaceKey is a property of TaskConfig this is namespace of the resource.
+	namespaceKey string = "namespace"
+	// kindKey is a property of TaskConfig this is kind of the resource.
+	kindKey string = "kind"
+)
+
+// CASTEngineBuilder helps to build a new instance of castEngine
+type CASTEngineBuilder struct {
+	RuntimeConfig []apis.Config
+	CASTemplate   *apis.CASTemplate
+	UnitOfUpgrade *upgrade.ResourceDetails
+	errors        []error
+}
+
+// String implements Stringer interface
+func (ceb CASTEngineBuilder) String() string {
+	return stringer.Yaml("cast engine builder", ceb)
+}
+
+// GoString implements GoStringer interface
+func (ceb CASTEngineBuilder) GoString() string {
+	return ceb.String()
+}
+
+// WithRuntimeConfig sets runtime config in CASTEngineBuilder.
+func (ceb *CASTEngineBuilder) WithRuntimeConfig(configs []upgrade.DataItem) *CASTEngineBuilder {
+	runtimeConfig := []apis.Config{}
+	for _, config := range configs {
+		c := apis.Config{
+			Name:    config.Name,
+			Value:   config.Value,
+			Data:    config.Entries,
+			Enabled: "true",
+		}
+		runtimeConfig = append(runtimeConfig, c)
+	}
+	ceb.RuntimeConfig = runtimeConfig
+	return ceb
+}
+
+// WithUnitOfUpgrade sets unitOfUpgrade details in CASTEngineBuilder.
+func (ceb *CASTEngineBuilder) WithUnitOfUpgrade(item *upgrade.ResourceDetails) *CASTEngineBuilder {
+	ceb.UnitOfUpgrade = item
+	return ceb
+}
+
+// WithCASTemplate sets casTemplate object in CASTEngineBuilder.
+func (ceb *CASTEngineBuilder) WithCASTemplate(casTemplate *apis.CASTemplate) *CASTEngineBuilder {
+	ceb.CASTemplate = casTemplate
+	return ceb
+}
+
+// validate validates CASTEngineBuilder struct.
+func (ceb *CASTEngineBuilder) validate() error {
+	if ceb.CASTemplate == nil {
+		return errors.New("missing castemplate")
+	}
+	if ceb.UnitOfUpgrade == nil {
+		return errors.New("missing upgrade item")
```

Comment:
we shall fix this in next PR
let aggregate all the validation specific errors together & return a formatted version

Reply:
TODO